### PR TITLE
Arrange class hierarchy for Inner helpers

### PR DESCRIPTION
**Pull request by broadwaylamb**
What's in this PR:

- when implementing helper subscription classes (`Inner`) for operators, there was a lot of code duplication. I attempted to extract the common pieces into base classes (@spadafiva you may want to take a look)
- tweaked the `DropWhile` implementation to match Apple's Combine behavior when requesting elements using a `Subscription` instance

### CODE REVIEW

File: Sources/OpenCombine/Publishers/OperatorSubscription.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/OperatorSubscription.swift:11]: I'm looking forward to tweaking my PRs to use this instead of what I've currently done in my other open ones.  I'm wondering if this should conform to `Subscription` also.  It seems like every subclass is saving the upstream subscription and then sending to the downstream.  We could the override that in Operators like `DropWhile` to request 1 instead of a different number.  Do you think there is enough meaningful overlap in the operators to do that?

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/Publishers.DropWhile.swift]: Maybe something like `DropPredicate` would be a more clear name?

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/Publishers.DropWhile.swift:127]: It's a little unclear to me what's happening here and why.  It could be I don't understand how demand works totally yet.  Is it worth leaving a comment explaining why you don't pass demand on to the upstream when the predicate is `nil`?

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: maybe `MapTransform` ?

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: If `OperatorSubscription` conformed to subscription, we would not have to have this receive here in the Map.Inner.

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (spadafiva) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: It doesn't look like you're transforming the completion here.  Can this just be `downstream.receive(completion: completion)` instead of the switch?

**Review by spadafiva — COMMENTED** [review]
@broadwaylamb Just had a few comments / questions on this one, but overall I'm looking forward to seeing it merged so that I can pull the changes into my PR branches.  It will definitely make implementing the operators a bit easier.

File: Sources/OpenCombine/Publishers/OperatorSubscription.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/OperatorSubscription.swift:11]: >Do you think there is enough meaningful overlap in the operators to do that?

I don't think so. This makes sense only for map-like operators, we don't have many of them.

I intentionally haven't provided any implementation of the `Subscription` protocol here to encourage implementors (including myself in the future) to really think about the semantics of the `request(_:)` method for every particular operator.

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.DropWhile.swift]: IMO adding the "Drop" word won't bring any new information to the reader — we're already dealing with DropWhile.

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.Map.swift]: No, the types won't match: the `completion` argument has the type `Subscribers.Completion<Upstream.Failure>`, and the `downstream.receive(completion:)` method expects `Subscribers.Completion<Error>`.

Since user-defined generic types in Swift are invariant relative to their generic parameters,  `Subscribers.Completion<Upstream.Failure>` is not a subtype of `Subscribers.Completion<Error>` and therefore cannot be passed where the latter is expected

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.DropWhile.swift:127]: Yeah, it's not obvious at first glance. Will add a comment.

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (spadafiva) [reply Sources/OpenCombine/Publishers/Publishers.Map.swift]: I see, that's a bit annoying that we can see it should be that way but that the type system can't.

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (spadafiva) [reply Sources/OpenCombine/Publishers/Publishers.DropWhile.swift]: That's fair enough.  I think I'm just finding the wording `Generic` confusing because I'm expecting it to be more encompassing than a local typealias.

**Comment by broadwaylamb** [comment]
@spadafiva I've addressed your feedback, can you take a look?

**Review by spadafiva — APPROVED** [review]
@broadwaylamb Looks good to me.  I think it's a lot clearer to understand and easier to read.

**Comment by spadafiva** [comment]
@broadwaylamb Once you've got it merged in, ping me if you remember so that I can update my PRs based on the changes from this one.

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (M0rtyMerr) [inline Sources/OpenCombine/Publishers/Publishers.DropWhile.swift:55]: Hey, great work :)
Could you please make these changes in a separate PR? These changes don't have a relation to `tryMap` operator

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (M0rtyMerr) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: Hey, great work :)
Why these reformatting changes were made? As I can see code fits swiftlint rules even without these changes. Many unnecessary changes which have no relation to feature (tryMap) complicate code review process. Could u please revert these changes?

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (M0rtyMerr) [inline Sources/OpenCombine/Publishers/Publishers.Map.swift]: What's the reason of moving CustomStringConvertible into class instead of leaving it in the extension?

File: Sources/OpenCombine/Publishers/OperatorSubscription.swift
Comment (MortyMerr) [reply Sources/OpenCombine/Publishers/OperatorSubscription.swift]: Agreed with @broadwaylamb.

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.Map.swift]: I've accidentally configured SwiftLint in such a way that comments were allowed to ignore the line length limit. Fixed that in master.

File: Sources/OpenCombine/Publishers/Publishers.Map.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.Map.swift]: I have a habit of implementing protocol requirements inside a class body instead of an extension if the class is non-final — this leaves the opportunity to override those methods if we need. Yep, in this case we probably won't need to override 'em, but this is a matter of consistency.

File: Sources/OpenCombine/Publishers/Publishers.DropWhile.swift
Comment (broadwaylamb) [reply Sources/OpenCombine/Publishers/Publishers.DropWhile.swift:55]: This PR is now about `Inner` refactoring, for `TryMap` see #15.

#15 depends on this PR, so it's better to merge this first, and review #15 after that.

**Comment by spadafiva** [comment]
@MortyMerr Can you take a look at this so the @broadwaylamb can merge?  I'm holding off on updating some of my PRs until this is resolved so that I don't have to refactor them again later.

**Review by M0rtyMerr — APPROVED** [review]

**Pull request merged**
